### DISCUSSION TITLE
Most efficient way to loop through a Mat

### DISCUSSION THREAD

**Discussion by msmartin4470** [rainyl/opencv_dart#44] [General]
What is the best way to loop though a Mat in dart?  In C I did this using pointers as follows.  Thanks in advance.

```
for (int i = 0; i < mat.rows; i++)
{
	uint16_t* matRow = mat.ptr<uint16_t>(i);
	for (int j = 0; j < mat.cols; j++)
	{
		uint16_t value = matRow[j];
		matRow[j] = lut.at<uint16_t>(value);
	}
}

```

**Comment by rainyl** [comment]
UPDATE: Since `v1.2.0`, Mat.at, Mat.set, Mat.ptrAt, etc. had been improved significantly, refer to #193 for more info.

@msmartin4470 You can use [Mat.at\<T\>()](https://pub.dev/documentation/opencv_dart/latest/opencv_dart/Mat/at.html) too, but only for getter, if you need setter, use [Mat.set\<T\>](https://pub.dev/documentation/opencv_dart/latest/opencv_dart/Mat/set.html)

Although not tested, I think it is an efficient way to loop through a mat for now.

**Reply by msmartin4470** [reply]
It seems that using at() and set() takes quite a bit more time that my existing solution which it implemented in c++ can called through a method channel.  I will work with the new ptr functions to see if it improves the speed.

**Reply by rainyl** [reply]
Your c++ implementation copies fewer times between dart side and c++ side, so it's unfair.

It is forseeable that at() and set() will be slow, I don't know your test code but it also surprised me, dart do not support functions overloading so `if else` is used to determine the type internally, https://github.com/rainyl/opencv_dart/blob/7e7a3df7be65e761ca403645d9f39113a2b98b87/lib/src/core/mat.dart#L240-L440 maybe we should just expose internal functions like `atU8`, `atF64`... `atVec3b`, etc., but it will be laborious... 

I will test the overhead of type determination when having some time, and try to improve the performance if the overhead is unacceptable.

**Comment by rainyl** [comment]
@msmartin4470 Recently I have implemented the Mat::ptr in v1.0.3 and v2.0.0-dev.4 to allow accessing pointer like c++, with proper and careful operation this might be more efficient, you can try it and open new issues if there are any problems.

Mat::ptr\<T\>() -> Mat.ptrAt\<T\>()

or

Mat::ptr\<uchar\>() -> Mat.ptrU8()

and more similar APIs, you can explore more in release and document.

**Reply by msmartin4470** [reply]
Thank you I will give this a try

**Reply by msmartin4470** [reply]
Do you have any quick thoughts on how to speed this up?  It is just applying a lookup table to a 16 bit image.  It is taking about 1500ms currently, and it's the one piece of my project that I need up run as fast as possible.  Thanks.

```
    final windowLeveled = src.clone();
    for (int r = 0; r < src.rows; r++) {
      final srcRow = src.ptrU16(r);
      final windowLeveledRow = windowLeveled.ptrU16(r);
      for (int c = 0; c < src.cols; c++) {
        final value = lut[srcRow[c]];
        windowLeveledRow[c] = value;
      }
    }
```

**Reply by rainyl** [reply]
https://docs.opencv.org/4.x/db/da5/tutorial_how_to_scan_images.html#:~:text=Goal

I think cv.LUT() is what you need, but it seems the api was not implemented in opencv_dart, if you need it urgently, you can try to add it and open a PR, or I will add it recently.

Actually it will be easy, the API exists in C wrappers but just not added to dart implementation, even no need to recompile the dynamic library.

**Reply by rainyl** [reply]
oh, sorry it seems LUT is only for CV_8U/S, so maybe we need to implement a more universal function like LUT

**Reply by msmartin4470** [reply]
Yes LUT is exactly what I need.  As you said it does not support CV_16U, so I am a bit stuck.  Thank you for looking into it.

**Reply by rainyl** [reply]
@msmartin4470 I have published v1.0.4 with cv.LUT supported, please upgrade and test if it works and works as expected.
